src/audio/mux_generic.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright (c) 2019, Intel Corporation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Artur Kloniecki <arturx.kloniecki@linux.intel.com>
+ */
+
+#include <config.h>
+
+#if CONFIG_COMP_MUX
+
+#include "mux.h"
+
+/* \brief Demuxing 16 bit streams.
+ * \param[in,out] dev Demux base component device.
+ * \param[in,out] sink Destination buffer.
+ * \param[in,out] source Source buffer.
+ * \param[in] frames Number of frames to process.
+ * \param[in] data Parameters describing channel count and routing.
+ */
+static void demux_s16le(struct comp_dev *dev, struct comp_buffer *sink,
+			struct comp_buffer *source, uint32_t frames,
+			struct mux_stream_data *data)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int32_t sample;
+	int16_t *src;
+	int16_t *dst;
+	uint8_t i;
+	uint8_t out_ch;
+	uint8_t in_ch;
+
+	for (i = 0; i < frames; i++) {
+		for (out_ch = 0; out_ch < data->num_channels; out_ch++) {
+			sample = 0;
+
```

Comment:
data->mask [out_chan] should be checked here because if it's 0 it will fail for every test within th einner loop. I've raised this before and you have not fixed this.